proposed/non-primitive-constant-data.md

```diff
@@ -0,0 +1,163 @@
+
+## Current valid constant forms
+Currently accepted are 4 forms of constant
+- Integer/Float constants
+```csharp
+void Constants()
+{
+   int x = 3;
+   double y = 4.0;
+}
+```
+
+Which is represented in IL via the various ldc opcodes.
+```
+IL_0001: ldc.i4.s 3
+IL_0003: stloc.s 0
+IL_0005: ldc.r8 4.0
+IL_000E: stloc.s 1
+```
+
+- String constants
+```csharp
+void Constants()
+{
+    string z = "z";
+}
+```
+
+Which is represented in IL using the ldstr instruction and #US metadata table.
+```
+IL_0001: ldstr "z"
+```
+- Byte data constants
+```csharp
+void Constants()
+{
+    ReadOnlySpan<byte> byte_data = new ReadyOnlySpan<byte>(new byte[]{0,1,2,3});
+}
+```
+Which is represented in IL using an RVA static field on an anonymous type and the accessed via ldsflda to get its address.
+```
+IL_0001: ldloca.s 0
+IL_0003: ldsflda valuetype '<PrivateImplementationDetails>'/'__StaticArrayInitTypeSize=3' '<PrivateImplementationDetails>'::'0C7A623FD2BBC05B06423BE359E4021D36E721AD'
+IL_0008: ldc.i4.3
+IL_0009: call instance void valuetype [System.Memory]System.ReadOnlySpan`1<uint8>::.ctor(void*, int32)
+```
+
+
+- Primitive data for a dynamically constructed array
+```csharp
+void Constants()
+{
+    int[] data = new byte[]{0,1,2,4};
+}
+```
+
+Which is represented in IL using an RVA static field on an anonymous type which is used in conjunction with the System.Runtime.CompilerServices.RuntimeHelpers.InitializeArray function to dynamically create an array based on a set of data which was statically placed into the assembly.
+```
+IL_0001: ldc.i4.4
+IL_0002: newarr [mscorlib]System.Byte
+IL_0007: dup
+IL_0008: ldtoken field int32 '<PrivateImplementationDetails>'::'12DADA1FFF4D4787ADE3333147202C3B443E376F'
+IL_000d: call void [mscorlib]System.Runtime.CompilerServices.RuntimeHelpers::InitializeArray(class [mscorlib]System.Array, valuetype [mscorlib]System.RuntimeFieldHandle)
+IL_0012: stloc.0
+```
+## New valid constant forms
+
+The proposal is to provide a set of CompilerServices apis implemented as intrinsics (in most cases) which will allow the use of more complex constants. This api will accept as a byref argument a reference to data in a well defined type layout, and return a byref to data in the correct platform layout.
+
+### Definition of types which can be represented as constants
+- Must be a struct
+- The struct must have sequential layout
+- Without pointers of any form (No object reference, IntPtr, UIntPtr, pointer, function pointer)
+- All fields must be public, or the type must be a primitive type
```

Comment:
Why does `public` matter at all here? I would assume the type **must** be primitive not any *or* qualification.

Reply:
> Why does public matter at all here?

The field layout has to be public contract for this to work with versioning. The easiest way to guarantee it is to require all fields to be public. Consider the case where the struct is defined in one assembly and used in different assembly.

> "opaque" types, such as `Vector128<T>`

Types like `Vector128<T>` do not work with this in general. The swizzler would have to special case them.

Reply:
> The field layout has to be public contract for this to work with versioning

I don't believe so. It could work with fields of any accessibility, provided the user doesn't do anything that modifies the layout of the struct later.

At the language level, it would be possible for the user to specially attribute structs that they want `constant` to work with (which tells the compiler, "I am opting into this behavior, and I won't change the layout later"). If they were to change it later, then it would be a breaking change and would fail at runtime (the same as any number of other breaking changes that are possible for people to make).

Reply:
> At the language level, it would be possible for the user to specially attribute structs that they want constant to work with

I think it's going to be a requirement for these types to be marked in a special way. There are too many requirement,s most outlined in this document, that go against how `struct` are generated in general.

Reply:
> The field layout has to be public contract for this to work with versioning. 

Disagree. The only requirement is the field layout does not change, not that the field layout is public. The consumer of the type doesn't need to be aware of the field ordering so long as the compiler can see the entirety of the `struct`

Reply:
> The only requirement is the field layout does not change

Right. That makes the field layout public contract. E.g. The reference assemblies and the tooling around them would need to learn how to deal with it.

Reply:
> The reference assemblies and the tooling around them would need to learn how to deal with it.

It would be great if the tooling could just reuse the C# Ref Assembly feature (which already deals with fields, etc)... It would be conceivable, even for partial facades, if the reference definitions were pulled from the general `S.P.Corelib` reference assembly (as needed), rather than being manually recreated.

Reply:
> It would be great if the tooling could just reuse the C# Ref Assembly

I agree that it would be great, but last time I have checked the C# Ref Assembly feature was insufficient for what CoreFX needs.

> rather than being manually recreated

The manually maintaned public surface definition lets us ensure that we have same surface between platforms, and compatible surface betwen versions and different runtimes. The C# Ref Assembly feature does not have equivalent for this today.

Reply:
>  Generally, non-public fields are not considered to be part of the public contract of a type. I am aware that in some circumstances, the private details of a structure affect various behaviors of the C#, but we generally want to make public contracts impacted by private details.

Struct private fields are very much a part of the c# contract. I think every time people have tried to get cute about them not they get broken. I don't think I'd want to require public fields here in C#. Would detract a bit from the feature.

Reply:
> I agree that it would be great, but last time I have checked the C# Ref Assembly feature was insufficient for what CoreFX needs.

We should follow up again on this as I'm not sure what feature is missing at this point. There shouldn't be any gap at this point, other than changing the infrastructure for calling the compiler.